### DISCUSSION TITLE
sqlite+aiosqlite sometimes doesn't find rows just added

### DISCUSSION THREAD

**Discussion by helgridly** [sqlalchemy/sqlalchemy#6987] [Usage Questions]
My code looks something like:

1. Start an asyncio coro "worker1" using `asyncio.create_task`
2. worker1 adds a row to the database, then starts another task "worker2", passing the PK
3. worker2 attempts to read the row back out of the database using the PK

Occasionally, worker2 fails to find the row.

I've fabricated a minimal-as-I-can-get-it repro; there's a lot of scaffolding, but the relevant part is [here](https://gist.github.com/helgridly/e3f4a4df46b2576f3560fedec7b6e1f1#file-repro-py-L116).

Some notes on the code:

- It starts 100 worker1 instances, generating and attempting to save UUIDs to a sqlite in-memory database.
- worker1 also saves the same UUID as a file.
- Generating the UUID is wrapped in a function that will raise an exception 90% of the time, to simulate errors.
- The UUID-generating function is wrapped in a "retry on failure" decorator.
- worker2 reports when it fails to find the row and prints the list of UUIDs out at the end of program execution.
- worker2 *always* succeeds in finding the file version of the UUID, so it's definitely a database issue.
- The repro contains both a sync and an async implementation of the database side, toggleable with a bool near the top of the file.

I cannot reproduce the issue if:

- I remove the retry-on-failure behaviour; OR
- if I switch to SQLAlchemy's SyncClient by setting `use_async_db=False`

My mental model (which is clearly failing me here) is that once `await session.commit()` completes, a session opened after that should see the committed data. It seems possible that in all the jumping around between asyncio tasks I'm failing to `await` the right thing. From my reading online I understand that sqlite is funny in how it handles writes, so I'm wondering if it's also possible that random chance is scheduling the worker2 instance ahead of when the row write.

Can anyone help me get to the bottom of this? TIA.

Ubuntu 20.04, Python 3.9.6, SQLAlchemy 1.4.23

---

Update: I've added more logging (updated the gist) and as far as I can tell the worker instances are getting scheduled in the correct order:

```
sqla_events.inner: returned e7700733-8fef-4470-b4e8-059d7c2d2499 on retry #0
sqla_events.sess_ctx: opened session 28
sqla_events.worker1: adding e7700733-8fef-4470-b4e8-059d7c2d2499 in session 28
sqla_events.sess_ctx: committed session 28
sqla_events.sess_ctx: opened session 37
sqla_events.worker2: querying for e7700733-8fef-4470-b4e8-059d7c2d2499 on 37
```

This leaves possible answers as:
1. PEBKAC
2. sqlite doing something funky behind the scenes
3. something I can't even fathom

**Comment by CaselIT** [comment]
Hi,

Have you tried using a normal db instead of an in memory one?

**Reply by helgridly** [reply]
Can't seem to repro it with a file-based db rather than `:memory:`. Doesn't repro on postgres with `postgresql+asyncpg` either.

**Reply by CaselIT** [reply]
I'm not sure if there's something funny going on with the StaticPool that's used by aiosqlite when using an in-memory db.

Thoughts @zzzeek ?

**Reply by zzzeek** [reply]
it's not possible to run concurrent tasks against a plain SQLite "memory" database without additional options; a "memory" connection supports only one database connection at a time.  for a shared memory database you need to use regular `QueuePool` or `NullPool` indicating a process-local shared cache as:

    engine = create_engine("sqlite+pysqlite:///file::memory:?cache=shared&uri=true")
